packages/animations/lib/src/fade_through_transition.dart

```diff
@@ -0,0 +1,329 @@
+class FadeThroughPageTransitionsBuilder extends PageTransitionsBuilder {
+  /// Creates a [FadeThroughPageTransitionsBuilder].
+  const FadeThroughPageTransitionsBuilder();
+
+  @override
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return FadeThroughTransition(
+      animation: animation,
+      secondaryAnimation: secondaryAnimation,
+      child: child,
+    );
+  }
+}
+
+/// Defines a transition in which outgoing elements fade out, then incoming
+/// elements fade in and scale up.
+///
+/// The Fade through pattern provides a transition animation between UI elements
+/// that do not have a strong relationship to one another. As an example, the
+/// [BottomNavigationBar] may use this animation to transition the currently
+/// displayed content when a new [BottomNavigationBarItem] is selected.
+///
+/// Scale is only applied to incoming elements to emphasize new content over
+/// old.
+///
+/// Consider using [FadeThroughPageTransitionsBuilder] within a
+/// [PageTransitionsTheme] if you want to apply this kind of transition to
+/// [MaterialPageRoute] transitions within a Navigator. Or use this transition
+/// directly in a [PageTransitionSwitcher.transitionBuilder] to transition
+/// from one widget to another.
+///
+/// See also:
+///
+///  * [Fade-through](https://spec.googleplex.com/draft/mio-design/motion-new/the-motion-system.html#fade-through)
+///    in the Material Design spec.
+class FadeThroughTransition extends StatefulWidget {
+  /// Creates a [FadeThroughTransition].
+  ///
+  /// The [animation] and [secondaryAnimation] argument are required and must
+  /// not be null.
+  const FadeThroughTransition({
+    @required this.animation,
+    @required this.secondaryAnimation,
+    this.child,
+  })  : assert(animation != null),
+        assert(secondaryAnimation != null);
+
+  /// The animation that drives the [child]'s entrance and exit.
+  ///
+  /// See also:
+  ///
+  ///  * [TransitionRoute.animate], which is the value given to this property
+  ///    when the [FadeThroughTransition] is used as a page transition.
+  final Animation<double> animation;
+
+  /// The animation that transitions [child] when new content is pushed on top
+  /// of it.
+  ///
+  /// See also:
+  ///
+  ///  * [TransitionRoute.secondaryAnimation], which is the value given to this
+  //     property when the [FadeThroughTransition] is used as a page transition.
+  final Animation<double> secondaryAnimation;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// This widget will transition in and out as driven by [animation] and
+  /// [secondaryAnimation].
+  final Widget child;
+
+  @override
+  State<FadeThroughTransition> createState() => _FadeThroughTransitionState();
+}
+
+class _FadeThroughTransitionState extends State<FadeThroughTransition> {
+  AnimationStatus _effectiveAnimationStatus;
```

Comment:
Is the _effectiveAnimationStatus tech going to be employed by other transitions? Seems like it would be ... maybe factor it out in the future

Reply:
Yes, I think this will be useful for other transitions implemented by @shihaohong. At that point we should factor this out.